File: blackjack/forth/outcomes.fs
variable total
variable subtotal
variable m
variable d

\ Can be done using "create"
create deck 4 , 4 , 4 , 4 , 4 , 4 , 4 , 4 , 4 , 16 ,

: cards cells deck + ;
: inc cards 1 swap +! ;
: dec cards -1 swap +! ;

: partitions recursive \ subtotal on stack

  subtotal !

  10 0 do
    i cells deck + @ 0 > if

      i 1 subtotal @ + + total !
      
      total @ 21 < if
	1 m +!
	subtotal @ total @
        i dec
	partitions
        i inc
	subtotal !
      else
	total @ 21 = if
	  1 m +!
	  leave
	then
      then
    then
  loop
;

variable n
0 n !

: deal
  \ Dealer showing i
  10 0 do
    i dec
    0 m !
    10 0 do
      i dec
      1 i + partitions
      i inc
    loop
    m @ n +!
    ." Dealer " i . m ? cr
    i inc
  loop ;
  
deal
." Total " n ? cr

bye

File: blackjack/r-c-.c/outcomes.R

dyn.load("partitions.so")

partitions <- function(cards,subtotal) {
    out <- .C("partitions",
              cards=as.integer(cards),
              subtotal=as.integer(subtotal),
              m=as.integer(0))
    return(out$m)
}


deck = c(4,4,4,4,4,4,4,4,4,16)
d=0

for (i in 1:10) {
    # Dealer showing
    deck[i] = deck[i]-1
    p = 0
    for (j in 1:10) {
        deck[j] = deck[j]-1
        n = partitions(deck,j)
        deck[j] = deck[j]+1
        p = p+n
    }

    print(paste("Dealer showing ",i-1," partitions = ",p,sep=""))
    d = d+p
    deck[i] = deck[i]+1
}
print(paste("Total partitions = ",d,sep=""))
